### PR TITLE
Fix Issue 13856 - std.stdio.readln stomps arrays

### PR DESCRIPTION
**Pull request by rainers**
This avoid interference of readln with standard arrays.
It also avoids an allocation when passing a stack allocated buffer that is large enough to hold the line.

https://issues.dlang.org/show_bug.cgi?id=13856

### CODE REVIEW

**Comment by rainers** [comment]
> N.B. Merge conflicts / needs rebase.

Sorry for the delay, now rebased.

**Comment by andralex** [comment]
is this good to go?

**Comment by rainers** [comment]
It's ok if we can live with a small loss in performance for some uses of readln. See the benchmarks above, there are also improvements in other cases. I couldn't check the performance on linux/OSX though.

**Comment by schveiguy** [comment]
I think the fact that using a stack-allocated array to buffer for readln results in always allocation is enough to make this pull worth doing.

**Comment by acehreli** [comment]
Will this make it into 2.067? It is blocker for me and I think stomping should be a big problem for readln. Think security! What about the children? :p

File: std/stdio.d [hunk 1]
```diff
@@ -3846,14 +3846,84 @@ unittest
...
+        if (cap < buf.length)
+            cap = buf.length;
+        pos = 0;
+    }
+    @property char[] data()
+    {
+        // always shrink/extend the buffer to allow reuse in the next call
+        if (initcap > 0 || cap > buf.length)
+            assumeSafeAppend(buf.ptr[0..pos]);
+        return buf.ptr[0..pos];
```
Comment (schveiguy) [inline std/stdio.d]: Just a note here, I know that this type is only used here. But having `data` always call `assumeSafeAppend` could be expensive, there is no reason to call it more than once for the same data. But practically speaking, the code only calls `data` right before returning (and destroying the appender struct).

**Comment by schveiguy** [comment]
Just did another quick review, looks fine to me.

**Comment by andralex** [comment]
The solution seems quite complicated. What is the root of the problem? Is that the fact that readln() writes all the way up to GC.sizeof?

**Comment by rainers** [comment]
> Is that the fact that readln() writes all the way up to GC.sizeof?

Yes. The fix is a bit complicated because it is conservative in changing the API.

As discussed above it would be faster and safer if readlnImpl would never shrink the passed buffer and return a slice into it. That would make assumeUnique on the returned slice invalid, though, so appending to it is likely to produce an allocation.

**Comment by schveiguy** [comment]
As @rainers alluded to, the issue is that there is no communication from one call of readln to the next call, indicating the ownership of the array.

But the big fix I see here is that you could never use a stack buffer for readln (it would just malloc immediately, even if the line fit into the passed buffer). This is somewhat different problem than the title indicates, but it's fixed by this as well.

The solution is complex because readln is complex. It's a different implementation per OS, optimized for that OS's C runtime's available tools.

The correct fix would be to return both the line length *and* the buffer length, so the next call to readln kept that info. But that means an API change.

File: std/stdio.d [hunk 2]
```diff
@@ -4222,6 +4258,39 @@ private size_t readlnImpl(FILE* fps, ref char[] buf, dchar terminator = '\n')
...
+unittest
+{
+    auto deleteme = testFilename();
+    scope(exit) std.file.remove(deleteme);
+
+    std.file.write(deleteme, "abcd\n0123456789abcde\n1234\n");
+    File f = File(deleteme, "rb");
+
+    char[] ln = new char[2];
+    assert(ln.capacity > 5);
```
Comment (MartinNowak) [inline std/stdio.d]: We shouldn't assert any behavior of the array cache or the GC implementation.

**Comment by MartinNowak** [comment]
Here is what we could do.

1. Tell people to use `stdin.byLine()` instead of looping on `stdin.readln(buf)`.
2. Add a hack that caches the maximum buffer length in a static variable across multiple calls to `readln` to make the current interface fast and correct.

Anyone knows why this turned into a regression?
https://issues.dlang.org/show_bug.cgi?id=14310

**Comment by MartinNowak** [comment]
As usual the interactions of the GC, arrays, slices, stomp protection, and the array append cache are way too complex for anyone to work with.

**Comment by MartinNowak** [comment]
We're lacking a `byLine(buf)` to replace reason completely.

**Comment by MartinNowak** [comment]
> Will this make it into 2.067? It is blocker for me and I think stomping should be a big problem for readln. Think security! What about the children? :package: 

This used to be broken before and I no longer want to delay the release. As a workaround you can use byLine instead.
To fix this problem, we should make readln correct but inefficient, deprecate it, and add a byLine(buf) as replacement.

**Comment by schveiguy** [comment]
> To fix this problem, we should make readln correct but inefficient, deprecate it, and add a byLine(buf) as replacement.

No. `readln` serves a completely different need. `byLine` is not meant for reading one line.

I agree with changing the API however, as discussed earlier in this thread. Even `getline` as defined on Posix systems gets it right.

**Comment by MartinNowak** [comment]
> No. readln serves a completely different need. byLine is not meant for reading one line.

It's mainly the loop usage which causes us headache.

The obvious API that takes a buffer and returns a slice can't be overloaded with readln, any good proposals for an API?

**Comment by CyberShadow** [comment]
Not corrupting memory trumps being slower. Discussions of new APIs aside, I think this should go in as-is for now. InvalidMemoryOperationErrors are difficult to debug. Besides, we have an I/O API overhaul in front of us still (`std.io`).

**Comment by rainers** [comment]
Rebased and conflicts resolved. These are now the timings of the benchmark above on windows:
```
win32/master
readln1: 129 ms
readln2: 170 ms
byLine: 127 ms
551312 bytes used

win64/master
readln1: 601 ms
readln2: 652 ms
byLine: 599 ms
801664 bytes used

win32/fix
readln1: 120 ms
readln2: 128 ms
byLine: 105 ms
2128 bytes used

win64/fix
readln1: 287 ms
readln2: 288 ms
byLine: 265 ms
5648 bytes used
```
So it's faster for all tests now, probably due to recent changes to byLine.

**Comment by CyberShadow** [comment]
Great! @MartinNowak @schveiguy OK to merge?

File: std/stdio.d (hunk 2 above)
Comment (rainers) [inline std/stdio.d]: > We shouldn't assert any behavior of the array cache or the GC implementation.

A failing test reveals a bad interaction between the code here and the GC code, so it cannot be added as a unittest without these details , e.g. the "should not reallocate" optimization depends on a minimal capacity. The alternative would be to always pass the test if the initial assumption about the capacity is not correct.

**Comment by schveiguy** [comment]
What's up with all the makefile changes?

**Comment by CyberShadow** [comment]
Looks like a case of git's `core.autocrlf` converting the Unix line endings to CR LF. @rainers you will need to fix that and rebase.

**Comment by rainers** [comment]
> Looks like a case of git's core.autocrlf converting the Unix line endings to CR LF. @rainers you will need to fix that and rebase.

Arrgh, this keeps coming back again and again... Should be fixed now.

**Comment by schveiguy** [comment]
> So it's faster for all tests now, probably due to recent changes to byLine.

Yeah, there was some focus on that. Also, `assumeSafeAppend` had a degenerate case if the Block info wasn't in the cache yet.

I assume nothing has really changed since I last said LGTM, I'm not a big fan of how rebasing erases all the historical commits unless they were commented on. In any case, I think it still looks good.

File: std/stdio.d [hunk 3]
```diff
@@ -3793,13 +3794,82 @@ unittest
...
+        buf = b;
+        cap = initcap = buf.capacity;
+        if (cap < buf.length)
+            cap = buf.length;
+        pos = 0;
+    }
+    @property char[] data()
+    {
+        // always shrink/extend the buffer to allow reuse in the next call
+        if (initcap > 0 || cap > buf.length)
```
Comment (MartinNowak) [inline std/stdio.d]: Seems like you could replace `cap` and `initcap` with a boolean flag `reuseBuf`.

File: std/stdio.d [hunk 4]
```diff
@@ -4091,14 +4091,84 @@ unittest
...
+    }
+    void putbuf(char[] b)
+    {
+        reserve(b.length);
+        memcpy(buf.ptr + pos, b.ptr, b.length);
+        pos += b.length;
+    }
+    void putonly(char[] b)
+    {
+        // assume this is the only put call
```
Comment (MartinNowak) [inline std/stdio.d]: Does this mean `assert(pos == 0)`?

File: std/stdio.d [hunk 5]
```diff
@@ -4091,14 +4091,84 @@ unittest
...
 }
 
+// roll our own appender, but with "safe" arrays
+private struct ReadlnAppender
+{
+    import core.stdc.string;
+
+    char[] buf;
+    size_t cap;
+    size_t pos;
```
Comment (MartinNowak) [inline std/stdio.d]: Why do you need both `cap` and `buf.length`?

Comment (rainers) [inline std/stdio.d]: The original length of the buffer is needed in `data()` to detect whether the buffer was reallocated.

File: std/stdio.d (hunk 3 above)
Comment (rainers) [inline std/stdio.d]: > Seems like you could replace initcap with a bool.

True, but that would add an unnecessary `size_t` to `bool` conversion.

File: std/stdio.d (hunk 4 above)
Comment (rainers) [inline std/stdio.d]: Yes. I've added the assert.

**Comment by rainers** [comment]
> So it's faster for all tests now, probably due to recent changes to byLine.

As said above, I have tested this only on Windows. `readlnImpl` is versioned for other platforms, so it would be nice if someone could verify the benchmarks on these.

**Comment by MartinNowak** [comment]
Will checks this, should be fine though.

**Comment by MartinNowak** [comment]
It's a lot slower on linux.

```sh
readln1: 80 ms
readln2: 97 ms
byLine: 62 ms
```
vs.
```sh
readln1: 49 ms
readln2: 49 ms
byLine: 38 ms
```

**Comment by MartinNowak** [comment]
Because `_d_arraysetcapacity` is called for each line to obtain the buffer capactiy.

**Comment by MartinNowak** [comment]
We should rethink the API and consider other fixes before hacking around an inherent problem.

> > To fix this problem, we should make readln correct but inefficient, deprecate it, and add a byLine(buf) as replacement.

> No. readln serves a completely different need. byLine is not meant for reading one line.

Yes, and that's why readln doesn't need to be optimized for reusing the buffer, we should have `byLine(buf)` for that or even better implement `byLine()` with a `scope` malloc/realloc/free buffer.

**Comment by CyberShadow** [comment]
> We should rethink the API and consider other fixes before hacking around an inherent problem.

Please let's fix the memory corruption first.

**Comment by MartinNowak** [comment]
- this bug exists since ages, can't be that urgent
- we can't correctly fix readln, but this patch unnecessarily slows down byLine
- workaround is simple, dup the line before appending to it
- we recently improved readln performance, this eats up the gain
  readln is particularly important because it's one of the first things people try/benchmark and D was notoriously slower here than most scripting languages

**Comment by schveiguy** [comment]
> Yes, and that's why readln doesn't need to be optimized for reusing the buffer, we should have byLine(buf) for that or even better implement byLine() with a scope malloc/realloc/free buffer.

Improving byLine is fine, we probably have more flexibility here. The designer in me cringes that we have separate implementations for byLine and readln...

To @CyberShadow's point, if the slowdown is not fixable without fixing the array stomping, then array stomping should have priority. Perhaps fixing byLine will alleviate this issue.

**Comment by rainers** [comment]
Rebased and ReadlnAppender removed from the HAS_GETDELIM version aswell as the bad GC.sizeof optimization. This is probably slightly faster for byLine, but will add a few allocations for the readln1 benchmark.

File: std/stdio.d (hunk 2 above)
Comment (rainers) [inline std/stdio.d]: > We shouldn't assert any behavior of the array cache or the GC implementation.

I've commented the implementation specific assertions.

**Comment by schveiguy** [comment]
What is the status on this?

**Comment by CyberShadow** [comment]
> readln is particularly important because it's one of the first things people try/benchmark and D was notoriously slower here than most scripting languages

Well, is it better if the D benchmark corrupts memory and dies, then? What sort of impression would that make, that D is so buggy it can't even do something as basic as read a few lines without crashing?

People keep running into this bug again and again:
http://forum.dlang.org/post/mzcnpsbdmfaxatdsqudf@forum.dlang.org
https://issues.dlang.org/show_bug.cgi?id=14578

I don't claim to understand the problem in full, but I can't believe there is something to disagree on here when the choice is between performance and *not corrupting memory*.

**Comment by schveiguy** [comment]
> People keep running into this bug again and again:

To be fair, as much as I want this to go through, those issues should be fixed by latest head.

**Comment by rainers** [comment]
> What is the status on this?

I guess it needs someone with a linux system to verify the latest version does not reduce the performance of the benchmarks.

> To be fair, as much as I want this to go through, those issues should be fixed by latest head.

Not sure why these should be fixed, the changed allocation pattern in byLine will just hide most occurences.

**Comment by DmitryOlshansky** [comment]
This is what I see on Linux x64 (best of 3 runs):
```
before:
readln1: 59 ms
readln2: 58 ms
byLine: 45 ms

after:
readln1: 80 ms
readln2: 57 ms
byLine: 45 ms
```

**Comment by DmitryOlshansky** [comment]
If the question is should we keep bogus but faster version then I suggest to merge and get correct version even if it's less efficient in 1 out of 3 use cases.

**Comment by schveiguy** [comment]
> This is what I see on Linux x64

Note that `byLine` has been changed not to use `readln`, so it can be removed from the benchmarks.

**Comment by schveiguy** [comment]
> Not sure why these should be fixed, the changed allocation pattern in byLine will just hide most occurences.

I didn't notice this message before, can you explain this further @rainers?

**Comment by rainers** [comment]
> I didn't notice this message before, can you explain this further

ByLine still calls readln, see ByLine.popFront.

The recent change in byLine is that it remembers the original buffer and keeps it for the next call to readln if the returned line is shorter. Otherwise the returned larger buffer is used for the next call. This causes larger buffers being used, so chances of hitting the last bytes of the buffer are reduced.

Some versions of readlnImpl use standard array operations on the buffer, e.g. buf ~=, buf.length = or buf.dup, which causes the allocation block to also hold array length information. These are stomped when GC.sizeof is used in a later call to determine the length of the buffer.

**Comment by DmitryOlshansky** [comment]
So @rainers @schveiguy and @MartinNowak  are we good to go?

**Comment by rainers** [comment]
> So @rainers @schveiguy and @MartinNowak are we good to go?

No objections from my side ;-) I just removed bad the CRs...

**Comment by schveiguy** [comment]
> So @rainers @schveiguy and @MartinNowak are we good to go?

I am a little hesitant because the performance difference is so large (according to your numbers, it's about 33% slower), and the `readln1` use case is the most common that could be used (just passing in the same buffer). I don't know if it's fixable without adjusting the API as discussed.

On the other hand, it's all an implementation issue, which can be fixed later if possible. One thing I foresee happening is someone complains "what the hell happened to performance of readln" and this gets reverted knee-jerkily. For some reason people are allergic to bug fixes that cut performance these days. Perhaps fixing memory corruption is a good enough defense. In any case, my LGTM still stands. I think it's a worthwhile change.

**Comment by CyberShadow** [comment]
> Perhaps fixing memory corruption is a good enough defense.

Do we have a test case which reproduces the memory corruption problems on master? As I understand, most common occurrences of memory corruption have already been fixed?

**Comment by DmitryOlshansky** [comment]
> Perhaps fixing memory corruption is a good enough defense

Being fast and incorrect is wrothless. So I'm merging this.

**Comment by schveiguy** [comment]
> Do we have a test case which reproduces the memory corruption problems on master?

The test cases that failed for the original byLine should also fail for readln loop implemented in the same way as readln1 (as that is how byLine was originally implemented).

**Comment by DmitryOlshansky** [comment]
@schveiguy add a test case like that in a separate pull ? Then reverting this would regress on the auto-tester.

**Comment by schveiguy** [comment]
> add a test case like that in a separate pull ?

I think @rainers already added one.

**Comment by DmitryOlshansky** [comment]
> But I don't think that would stop a revert, they'd just revert the test too.

Accidental revert would be stopped. Reverting test case is an extra burden that won't go lightly.

**Comment by schveiguy** [comment]
> Accidental revert would be stopped. Reverting test case is an extra burden that won't go lightly.

I hope accidental revert would never happen.

I also have seen most of the time reverts do not happen lightly, and such a move (committing test case in separate pull to prevent reverts) wouldn't be viewed positively.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by rainers** [comment]
> Being fast and incorrect is worthless. So I'm merging this.

Thanks.

> those weren't memory corruption issues, but asserts. Memory corruption is quite easy, just use a slice of an existing buffer as the buffer for readln

The unittest in this PR shows some failing use patterns, but this depends on implementation details of rt.lifetime. The stomping initially doesn't corrupt any memory outside the memory block, but this can happen later by a combination of the BlkAttr.APPENDABLE flag and bad length information, for example when trying to append to the array. The actual effect is hard to predict and can be anything from going by undetected and the test application crashing.

**Comment by rainers** [comment]
Regarding the performance drop: maybe we should just add an example to readln that uses the same pattern as byLine:
```
    char[] buf;
    while(!f.eof())
    {
        char[] ln = buf;
        f.readln(ln);
        if (ln.length > buf.length)
            buf = ln;
        /// use ln
    }
 ```
This should yield the same performance as byLine.

**Comment by schveiguy** [comment]
> we should just add an example to readln that uses the same pattern

Yes, please!

**Comment by DmitryOlshansky** [comment]
@rainers Good idea - something along the lines of being roughly equivalent to this code.

**Comment by rainers** [comment]
https://github.com/D-Programming-Language/phobos/pull/3485

**Pull request merged**